### PR TITLE
[clang] Use new interpreter in EvaluateAsConstantExpr if requested

### PR DESCRIPTION
**Pull request by tbaederr**
EvaluateAsConstantExpr() uses ::EvaluateInPlace() directly, which does not use the new interpreter if requested. Do it here, which is the same pattern we use in EvaluateAsInitializer.

### CODE REVIEW

**Comment by tbaederr** [comment]
Bot complains about lines I haven't touched :shrug:

**Review by erichkeane — COMMENTED** [review]
I don't have a good idea how a release note could be phrased here, but if you could try one, it would be appreciated and I think it is worth doing.

Side note: it would be really cool if at the end of the release cycle, you did a section in the release-notes with a 'status' of the new interpreter.  Basically a 'things that don't work', or a 'percent of tests we have that work', etc?

**Comment by tbaederr** [comment]
> I don't have a good idea how a release note could be phrased here, but if you could try one, it would be appreciated and I think it is worth doing.

You mean about this change specifically?

> Side note: it would be really cool if at the end of the release cycle, you did a section in the release-notes with a 'status' of the new interpreter. Basically a 'things that don't work', or a 'percent of tests we have that work', etc?

I could force the new interpreter and run the test suite, but that would only be a very rough estimate.

**Comment by tbaederr** [comment]
Ping

File: clang/lib/AST/Interp/ByteCodeExprGen.h [hunk 1]
```diff
@@ -129,7 +129,13 @@ class ByteCodeExprGen : public ConstStmtVisitor<ByteCodeExprGen<Emitter>, bool>,
...
   /// Classifies a type.
   std::optional<PrimType> classify(const Expr *E) const {
-    return E->isGLValue() ? PT_Ptr : classify(E->getType());
+    if (E->isGLValue()) {
+      if (E->getType()->isFunctionType())
+        return PT_FnPtr;
+      return PT_Ptr;
+    }
+
+    return classify(E->getType());
```
Comment (AaronBallman) [inline clang/lib/AST/Interp/ByteCodeExprGen.h:138]: This feels unrelated to the patch title; can you explain it a bit more? (It doesn't seem wrong, just not seeing how it ties to the test behavior changes.)

Comment (tbaederr) [reply clang/lib/AST/Interp/ByteCodeExprGen.h:138]: This is basically just an oversight;  This should've been added when we added function pointers. This case was just never exercised. I can tell you what test case trips this up as soon as my local build finished.

Comment (AaronBallman) [reply clang/lib/AST/Interp/ByteCodeExprGen.h:138]: Ah, let's pull those changes from this PR and land them separately then.

**Review by AaronBallman — APPROVED** [review]
LGTM with unrelated changes pulled; I'm fine landing this PR after landing the fix for the function pointer classification changes if necessary.
